src/owr/OptimisticWithdrawalRecipientPectra.sol

```diff
@@ -0,0 +1,451 @@
+
+    (address principalRecipient, address rewardRecipient, uint256 amountOfPrincipalStake) = getTranches();
+
+    // determine which recipeint is getting paid based on funds to be
+    // distributed
+    uint256 _principalPayout = 0;
+    uint256 _rewardPayout = 0;
+
+    unchecked {
+      // _claimedPrincipalFunds should always be <= amountOfPrincipalStake
+      uint256 principalStakeRemaining = amountOfPrincipalStake - _claimedPrincipalFunds;
+
+      if (_fundsToBeDistributed >= BALANCE_CLASSIFICATION_THRESHOLD && principalStakeRemaining > 0) {
+        if (_fundsToBeDistributed > principalStakeRemaining) {
+          // this means there is reward part of the funds to be
+          // distributed
+          _principalPayout = principalStakeRemaining;
+          // shouldn't underflow
+          _rewardPayout = _fundsToBeDistributed - principalStakeRemaining;
+        } else {
+          // this means there is no reward part of the funds to be
+          // distributed
+          _principalPayout = _fundsToBeDistributed;
+        }
+      } else {
+        _rewardPayout = _fundsToBeDistributed;
+      }
+    }
+
+    {
+      if (_fundsToBeDistributed > type(uint128).max) revert InvalidDistribution_TooLarge();
+      // Write to storage
+      // the principal value
+      // it cannot overflow because _principalPayout < _fundsToBeDistributed
+      if (_principalPayout > 0) claimedPrincipalFunds += uint128(_principalPayout);
+    }
+
+    /// interactions
+
+    // pay outs
+    // earlier tranche recipients may try to re-enter but will cause fn to
+    // revert
+    // when later external calls fail (bc balance is emptied early)
+
+    // pay out principal
+    _payout(principalRecipient, _principalPayout, pullFlowFlag);
+    // pay out reward
+    _payout(rewardRecipient, _rewardPayout, pullFlowFlag);
+
+    if (pullFlowFlag == PULL) {
+      if (_principalPayout > 0 || _rewardPayout > 0) {
+        // Write to storage
+        fundsPendingWithdrawal = uint128(_memoryFundsPendingWithdrawal + _principalPayout + _rewardPayout);
+      }
+    }
+
+    emit DistributeFunds(_principalPayout, _rewardPayout, pullFlowFlag);
+  }
+
+  function _payout(address recipient, uint256 payoutAmount, uint256 pullFlowFlag) internal {
+    if (payoutAmount > 0) {
+      if (pullFlowFlag == PULL) {
+        // Write to Storage
+        pullBalances[recipient] += payoutAmount;
+      } else {
+        recipient.safeTransferETH(payoutAmount);
+      }
+    }
+  }
+
+  function _requestWithdrawal(bytes memory pubkey, uint8 amount, bool _rewards) private {
+    if (pubkey.length != 48) revert InvalidWithdrawal_Failed();
+
+    if (
+      (!_rewards && amount < BALANCE_CLASSIFICATION_THRESHOLD)
+        || (_rewards && amount >= BALANCE_CLASSIFICATION_THRESHOLD)
+    ) {
+      if (_rewards) revert InvalidPectraWithdrawal_Rewards();
+      else revert InvalidPectraWithdrawal_Principal();
+    }
```

Comment:
What are we trying to achieve here? I feel like this could be simplified, I have a hard time parsing it with an if(and or and)(if/else).